### ISSUE TITLE
Supernova: limit in OSC message size querying node tree

### ISSUE THREAD

**Issue by dyfer** [supercollider/supercollider#3269]
Hello,
I came across inconsistency between scsynth and supernova when querying the node tree. Scsynth responds to the "/g_queryTree" message with large OSC messages, while supernova reports

> unhandled exception while running callback: send_to: Message too long

when the reply message would exceed around 9250 bytes. 
It prevents from using this mechanism to track synth status in slightly more complex pieces.
This was tested on a recent development build on macos 10.12. Reproducer:
```
(
var maxsynth = 241; //OK - osc message size 9216 bytes
// var maxsynth = 242; //supernova reports message too long (when running scsynth, SC reports osc message size 9256)

Server.supernova;
// Server.scsynth; //for comparison

s.waitForBoot({
	var defName = \abcdefgh;

	SynthDef(defName, {arg amp = 0;
		var sig = PinkNoise.ar(1);
		Out.ar(0, sig * amp);
	}).send(s);

	p = ParGroup.new;

	fork{
		s.sync;
		"starting synths:".postln;
		maxsynth.do({
			Synth(defName, [\amp, 0], target: p);
			".".post;
			0.025.wait;
		});
		postf("\n% synths launched.\n", maxsynth);
		OSCFunc({arg msg;
			~msg = msg;
			msg.postcs;
			"number of elements: ".post; msg.size.postln;
			"OSC message size (bytes): ".post; msg.msgSize.postln;
		}, '/g_queryTree.reply').oneShot;
		s.sendMsg("/g_queryTree", 0, 1);
	}
})
)
```
This can also be seen with smaller number of synths, that have more arguments:
```
(
var maxsynth = 103; //OK - osc message size 9124 bytes
// var maxsynth = 104; //supernova reports message too long (when running scsynth, SC reports osc message size 10044)

Server.supernova;
// Server.scsynth; //for comparison

s.waitForBoot({
	var defName = \abcdefgh;

	SynthDef(defName, {arg amp = 0, out = 0, freq = 100, att = 0.1, rel = 2, buffer = 0;
		var sig = PinkNoise.ar(1);
		Out.ar(0, sig * amp);
	}).send(s);

	p = ParGroup.new;

	fork{
		s.sync;
		"starting synths:".postln;
		maxsynth.do({
			Synth(defName, [\amp, 0], target: p);
			".".post;
			0.025.wait;
		});
		postf("\n% synths launched.\n", maxsynth);
		OSCFunc({arg msg;
			~msg = msg;
			// msg.postcs;
			"number of elements: ".post; msg.size.postln;
			"OSC message size (bytes): ".post; msg.msgSize.postln;
		}, '/g_queryTree.reply').oneShot;
		s.sendMsg("/g_queryTree", 0, 1);
	}
})
)
```

**Comment by vivid-synth** [comment]
@timblechmann do you have any advice on how to address this one?

**Comment by brianlheim** [comment]
In both supernova and scsynth, the error happens here, at this call to [sendmsg](https://linux.die.net/man/2/sendmsg): https://github.com/supercollider/supercollider/blob/develop/external_libraries/boost/boost/asio/detail/impl/socket_ops.ipp#L1299

I set a breakpoint at that location and it seems like what's being passed in is pretty much identical in both cases, so I think it has something to do with how the sockets are created.

**Comment by brianlheim** [comment]
Boo, it looks like both are created with the same arguments too ([here](https://github.com/supercollider/supercollider/blob/develop/external_libraries/boost/boost/asio/detail/impl/socket_ops.ipp?utf8=%E2%9C%93#L1376)). I think it needs a closer look, I wasn't super precise the first time around.

**Comment by sonoro1234** [comment]
I dont get this behaviour with supernova on Lua2SC. Is still happening?. If it is still happening could it be a client issue?

**Comment by dyfer** [comment]
@sonoro1234 Yes, this is still happening. I don't think it can be a client issue, it is being reported by supernova, while processing a simple osc message ["/g_queryTree", 0, 1]. 
What's your platform? I'm on macOS 10.12, I might have also tested in on Linux but can't remember.

**Comment by jamshark70** [comment]
Does Lua2SC use UDP or TCP to connect?

**Comment by sonoro1234** [comment]
it can use both but the test was with TCP. My platform is windows.

**Comment by jamshark70** [comment]
I'm not sure if TCP can avoid the problem. But: the supernova executable is the same codebase and the OSC request is the same. Only the client is different. Two different clients, both using UDP, should not cause different behavior in the server. But if one is using TCP instead, that could cause differences, particularly when the issue involves packet sizes (where TCP is more robust). So I'm guessing that might be why you couldn't reproduce it. (But it could be another reason. My thinking here is circumstantial, or "educated guess," well short of proven.)

**Comment by scztt** [comment]
I just ran into this for the first time over the weekend (synths with lots of envelope parameters fill up the OSC message quickly). Good to know it doesn't repro with TCP - though it would still be good to refactor `g_queryTree` to return results over multiple packets for complex cases...

**Comment by vivid-synth** [comment]
@sonoro1234 this is getting somewhat off topic but you experienced lost packets in practice? Can you describe what you were doing that caused them?

**Comment by sonoro1234** [comment]
@vivid-synth sending alot of data to create and update some synths that provide data in control buses to make my virtual humans talk.
With udp the behaviour was to randomly loose some packets so that virtual human talk was broken in some places. All was solved switching to tcp.

**Comment by telephon** [comment]
if there is a way to do it via udp we should do it via udp. It wouldn't be such a good idea to make tcp mandatory.
